### PR TITLE
Stable Diffusion CUDA Optimizations Part 3

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description

The third part for stable diffusion CUDA optimizations (https://github.com/microsoft/onnxruntime/pull/14428).

Included:
(1) Add options --force_fp32_ops, --inspect in optimize script so that user could force some operator to run in float32 to potentially get better image quality (with cost of performance).
(2) Update float16 conversion to handle Resize and GroupNorm. This could reduce two Cast nodes for each Resize op in fp16 model.
(3) Force inputs and outputs to be float16 to avoid data casts in the pipeline.
(4) Add Attention fusion for VAE decoder.
(5) Add BiasAdd operator to replace two Add (bias and residual); Add fusion for BiasAdd 

Performance tests show slight improvement in T4. Average latency reduced 0.1 seconds (from 5.35s to 5.25s) for 512x512 in 50 steps.

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/diffusion/bias_add_impl.cu [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  int32_t base_offset = blockIdx.x * C + threadIdx.x;
+  int32_t bias_offset = threadIdx.x;
+
+#pragma unroll
+  for (int32_t i = 0; i < C / TPB; ++i) {
+#if defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 530)
+    output[base_offset] = input[base_offset] + bias[bias_offset] + residual[base_offset];
+#else
+    output[base_offset] = static_cast<T>(float(input[base_offset]) + float(bias[bias_offset]) + float(residual[base_offset]));
+#endif
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/diffusion/bias_add_impl.cu:41]: include cu_inc/common.cuh?

Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/diffusion/bias_add_impl.cu:41]: Included (see above) but the amd build still failed so I have to add this.

Comment (yufenglee) [reply onnxruntime/contrib_ops/cuda/diffusion/bias_add_impl.cu:41]: oh, i see. you need to add 'using namespace onnxruntime::cuda;' to make compiler know the override operators

File: onnxruntime/contrib_ops/cuda/diffusion/bias_add.cc [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+using namespace ONNX_NAMESPACE;
+
+template <typename T>
+BiasAdd<T>::BiasAdd(const OpKernelInfo& op_info) : CudaKernel(op_info) {
+}
+
+template <typename T>
+Status BiasAdd<T>::ComputeInternal(OpKernelContext* context) const {
+  // Input:  [batch_size, height*width, channels]
+  // Bias:   [hidden_size]
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/diffusion/bias_add.cc:35]: [channels] ?

File: onnxruntime/python/tools/transformers/fusion_attention_vae.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        k_bias = numpy_helper.to_array(k_bias_tensor)
+        v_bias = numpy_helper.to_array(v_bias_tensor)
+
+        q_bias_shape = np.prod(q_bias.shape)
+        k_bias_shape = np.prod(k_bias.shape)
+        v_bias_shape = np.prod(v_bias.shape)
+
+        # Sometimes weights are stored in fp16
+        if q_weight_tensor.data_type == 10:
+            logger.debug("weights are in fp16. Please run fp16 conversion after optimization")
```
Comment (wangyems) [inline onnxruntime/python/tools/transformers/fusion_attention_vae.py:134]: we can do this (https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/fusion_attention.py#L324) if the model is exported fp16

**Review by wangyems — DISMISSED** [review]

File: onnxruntime/contrib_ops/cuda/diffusion/bias_add.cc (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/diffusion/bias_add.cc]: Good catch. Fixed it.

File: onnxruntime/python/tools/transformers/fusion_attention_vae.py (hunk 3 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/fusion_attention_vae.py:134]: Good tips. Right now this fusion only support fp32 model inputs, and this is a sanity check.

**Review by wangyems — APPROVED** [review]

**Pull request merged**